src/names/applications.cpp

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2021 Jeremy Rand
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <names/applications.h>
+
+#include <names/encoding.h>
+
+#include <regex>
+
+namespace
+{
+
+NameNamespace
+NamespaceFromString (const std::string& str)
+{
+    if (str == "d/")
+        return NameNamespace::Domain;
+    if (str == "dd/")
+        return NameNamespace::DomainData;
+    if (str == "id/")
+        return NameNamespace::Identity;
+    if (str == "idd/")
+        return NameNamespace::IdentityData;
+
+    return NameNamespace::NonStandard;
+}
+
+std::string
+NamespaceToString (NameNamespace ns)
+{
+    switch (ns)
+    {
+        case NameNamespace::Domain:
+            return "d/";
+        case NameNamespace::DomainData:
+            return "dd/";
+        case NameNamespace::Identity:
+            return "id/";
+        case NameNamespace::IdentityData:
+            return "idd/";
+        case NameNamespace::NonStandard:
+            return "";
+    }
+
+    /* NameNamespace values are only ever created internally in the binary
+    and not received externally (except as string).  Thus it should never
+    happen (and if it does, is a severe bug) that we see an unexpected
+    value.  */
+    assert (false);
+}
+
+NameNamespace
+PurportedNamespaceFromName (const std::string& name)
+{
+    const auto slashPos = name.find("/");
+    if (slashPos == std::string::npos)
+    {
+        return NameNamespace::NonStandard;
+    }
+
+    const std::string namespaceStr = name.substr(0, slashPos+1);
+    return NamespaceFromString(namespaceStr);
+}
+
+} // anonymous namespace
+
+NameNamespace
+NamespaceFromName (const std::string& name)
+{
+    const NameNamespace purported = PurportedNamespaceFromName(name);
+    const std::string purportedStr = NamespaceToString(purported);
+    const auto purportedLen = purportedStr.length();
+
+    const std::string label = name.substr(purportedLen);
+
+    if (label.length() < 1)
```

Comment:
I would suggest using `label.empty()` instead, which seems a bit more expressive to me.

Reply:
Thanks, will fix.